lib/mlir/Dialect/Clift/Utils/ImportModel.cpp

```diff
@@ -0,0 +1,585 @@
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::EnumType: {
+      RequireComplete = true;
+      const auto &T = llvm::cast<model::EnumType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      const auto &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      const auto &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      const auto &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      const auto &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+
+    case model::TypeKind::PrimitiveType:
+      revng_abort("Primitive types have no corresponding attribute.");
+
+    case model::TypeKind::Invalid:
+    case model::TypeKind::Count:
+      revng_abort("These are invalid values. Something has gone wrong.");
+    }
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getUnwrappedValueType(const model::Type &ModelType,
+                        bool RequireComplete = false,
+                        const bool Const = false) {
+    if (const auto *const T = llvm::dyn_cast<model::PrimitiveType>(&ModelType))
+      rc_return getPrimitiveType(*T, Const);
+
+    const auto getDefinedType = [&](const auto Attr) -> clift::ValueType {
+      return make<clift::DefinedType>(Attr, getBool(Const));
+    };
+
+    if (const auto It = Cache.find(ModelType.ID()); It != Cache.end())
+      rc_return getDefinedType(It->second);
+
+    const clift::TypeDefinition Attr = getTypeAttribute(ModelType,
+                                                        RequireComplete);
+
+    if (not Attr)
+      rc_return nullptr;
+
+    if (RequireComplete) {
+      const auto R = Cache.try_emplace(ModelType.ID(), Attr);
+      revng_assert(R.second);
+    }
+
+    rc_return getDefinedType(Attr);
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedValueType(const model::QualifiedType &ModelType,
+                        bool RequireComplete = false) {
+    if (not ModelType.UnqualifiedType().isValid()) {
+      if (EmitError)
+        EmitError() << "Invalid UnqualifiedType in QualifiedType";
+      rc_return nullptr;
+    }
+
+    auto Qualifiers = llvm::ArrayRef(ModelType.Qualifiers());
+
+    // Qualifier::isPointer verifies itself, which interferes with unit testing.
+    static constexpr auto IsPointer = [](const model::Qualifier &Q) {
+      return Q.Kind() == model::QualifierKind::Pointer;
+    };
+
+    // If the set of qualifiers contains any pointers,
+    // the base type does not need to be complete.
+    if (RequireComplete and llvm::any_of(Qualifiers, IsPointer))
```

Comment:
Now that I think about it, we have `model::Qualifier::isPointer` instead of `IsPointer`.

We also have `mode::Qualifier::isConst` instead of `IsConst` below.

Reply:
Forget about this. I finally understood what you meant in you other comment.